lightning/src/routing/onion_message.rs

```diff
@@ -0,0 +1,328 @@
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Onion message pathfinding lives here.
+
+use bitcoin::secp256k1::{self, PublicKey};
+
+use ln::channelmanager::ChannelDetails;
+use routing::gossip::{NetworkGraph, NodeId};
+use util::logger::Logger;
+
+use alloc::collections::BinaryHeap;
+use core::cmp;
+use core::ops::Deref;
+use prelude::*;
+
+/// Finds a route from us to the given `destination` node.
+/// If we have private channels, it may be useful to fill in `first_hops` with the results from
+/// [`ChannelManager::list_usable_channels`]. If `first_hops` is not filled in, connected peers
+/// without public channels will not be forwarded over.
+///
+/// [`ChannelManager::list_usable_channels`]: crate::ln::channelmanager::ChannelManager::list_usable_channels
+pub fn find_path<L: Deref, GL: Deref>(
+	our_node_pubkey: &PublicKey, destination: &PublicKey, network_graph: &NetworkGraph<GL>, first_hops: Option<&[&ChannelDetails]>, logger: L
+) -> Result<Vec<PublicKey>, Error> where L::Target: Logger, GL::Target: Logger
+{
+	log_trace!(logger, "Searching for an onion message path from origin {} to destination {} and {} first hops {}overriding the network graph",
+		our_node_pubkey, destination, first_hops.map(|hops| hops.len()).unwrap_or(0),
+		if first_hops.is_some() { "" } else { "not " });
+	let graph_lock = network_graph.read_only();
+	let network_channels = graph_lock.channels();
+	let network_nodes = graph_lock.nodes();
+	let our_node_id = NodeId::from_pubkey(our_node_pubkey);
+	let dest_node_id = NodeId::from_pubkey(destination);
+
+	// Add our start and first-hops to `frontier`.
+	let start = NodeId::from_pubkey(&our_node_pubkey);
+	let mut valid_first_hops = HashSet::new();
+	let mut frontier = BinaryHeap::new();
+	frontier.push(PathBuildingHop { cost: 0, node_id: start, parent_node_id: start });
+	if let Some(first_hops) = first_hops {
+		for hop in first_hops {
+			if !hop.counterparty.features.supports_onion_messages() { continue; }
+			let node_id = NodeId::from_pubkey(&hop.counterparty.node_id);
+			frontier.push(PathBuildingHop { cost: 1, node_id, parent_node_id: start });
+			valid_first_hops.insert(node_id);
+		}
+	}
+
+	let mut visited = HashMap::new();
+	while let Some(PathBuildingHop { cost, node_id, parent_node_id }) = frontier.pop() {
+		match visited.entry(node_id) {
+			hash_map::Entry::Occupied(_) => continue,
+			hash_map::Entry::Vacant(e) => e.insert(parent_node_id),
+		};
+		if node_id == dest_node_id {
+			return Ok(reverse_path(visited, our_node_id, dest_node_id, logger)?)
+		}
+		if let Some(node_info) = network_nodes.get(&node_id) {
+			// Only consider the network graph if first_hops does not override it.
+			if valid_first_hops.contains(&node_id) || node_id == our_node_id {
+			} else if let Some(node_ann) = &node_info.announcement_info {
```

Comment:
IIUC this is only to account for the case where the first hop features are different than the network-announced ones? What would be a case in which we can expect that?

Also, this currently seems to default to `supports_onion_messages == true` in case we haven't gotten a node announcement. Maybe this should rather be pessimistic and exclude a hop that may not support onion messages?

Reply:
> Also, this currently seems to default to supports_onion_messages == true in case we haven't gotten a node announcement. Maybe this should rather be pessimistic and exclude a hop that may not support onion messages?

Indeed, that was a regression 🙈 

> IIUC this is only to account for the case where the first hop features are different than the network-announced ones? What would be a case in which we can expect that?

Basically our local channel view is expected to be the most up-to-date. Gossip can be slow but we'll be the first to know about feature upgrade/downgrades for our channel peers

Reply:
> Basically our local channel view is expected to be the most up-to-date. Gossip can be slow but we'll be the first to know about feature upgrade/downgrades for our channel peers

Ah, I see. Makes sense!